### PR TITLE
feat(cli): implement ask for task run

### PR DESCRIPTION
**Pull request by Louise15926**
This PR has two major changes:  1. add ask to task run;  2. add a required flag for task-family name, update validate and ask correspondingly 

Related #702 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: internal/pkg/cli/task_run.go [hunk 1]
```diff
@@ -81,6 +87,12 @@ func (o *runTaskOpts) Validate() error {
 		return errMemNotPositive
 	}
 
+	if o.familyName != "" {
+		if err := validateTaskFamilyName(o.familyName); err != nil {
```
[the remark this answers is not in the document]
Comment (Louise15926) [reply internal/pkg/cli/task_run.go:91]: My thought was that it might be better to provide earlier feedback on naming than later. Since app and svc validate their names during validate and ask, I thought it could also be the case for task.

Since this flag was added after the design meeting (sorryy about the confusion😥), here is a little explanation. The field `familyName` is to task as `AppName` is to app or as `Name` is to svc. It will be later used to name not only task definition, but also the cf stack, log group and the ecr repo. So it needs to be an intersection of the set of valid names for each of these resources.

[the remark this answers is not in the document]
Comment (iamhopaul123) [reply internal/pkg/cli/task_run.go:91]: lol it's just a small nit i think but if we decide to keep this validation maybe we should allow underscore? https://github.com/aws/amazon-ecs-cli-v2/pull/1036/files#diff-cf1619746faab25a67237648f6e3594aL141

[the remark this answers is not in the document]
Comment (Louise15926) [reply internal/pkg/cli/task_run.go:91]: cfn stack name doesn't seem to accept underscore though. If we allow underscore then the name could be potentially invalid for stack?

[the remark this answers is not in the document]
Comment (iamhopaul123) [reply internal/pkg/cli/task_run.go:91]: i don't think so. I tried to deploy a CFN template with underscore in my task def name and it worked well. Any reference for the reason why you think it doesn't accept underscore as value (maybe you mean the logicalID doesn't support underscore)? Actually using underscore in family name is pretty common I think :P

[the remark this answers is not in the document]
Comment (iamhopaul123) [reply internal/pkg/cli/task_run.go:91]: oh i get you. it's the stack name. Then it makes sense to me!

File: internal/pkg/cli/task_run.go [hunk 2]
```diff
@@ -141,8 +219,8 @@ func BuildTaskRunCmd() *cobra.Command {
 		Example: `
 Run a task with default setting.
 /code $ copilot task run
-Run a task in the "test" environment under the current workspace.
-/code $ copilot task run --env test
+Run a task named "db-migrate" in the "test" environment under the current workspace.
+/code $ copilot task run -n db-migrate --env test
```
Comment (pkandasamy91) [inline internal/pkg/cli/task_run.go:223]: We might want to rethink about using -n as the flag (to me -n makes more sense for num) Would it make sense to name it family-name instead?

File: internal/pkg/cli/flag.go [hunk 3]
```diff
@@ -147,6 +147,7 @@ Must be of the format '<name>:<dataType>'. Can be specified multiple times.`
 	securityGroupsFlagDescription = "Optional. The security group id(s) for the task to use. Can be specified multiple times."
 	envVarsFlagDescription        = "Optional. Environment variables specified by key=value separated with commas."
 	commandsFlagDescription       = "Optional. List of commands that are passed to docker run. Can be specified multiple times."
+	taskFamilyFlagDescription     = "Required. The family name of the task. Tasks with the same family name share the same set of resources."
```
Comment (pkandasamy91) [inline internal/pkg/cli/flag.go:150]: Not sure that you need to mention that this is Required (since it's not optional)

File: internal/pkg/cli/task_run_test.go [hunk 4]
```diff
@@ -169,7 +175,7 @@ func TestTaskRunOpts_Validate(t *testing.T) {
 			basicOpts: defaultOpts,
 
 			appName: "my-app",
-			inEnv: "dev",
+			inEnv:   "dev",
```
Comment (pkandasamy91) [inline internal/pkg/cli/task_run_test.go:178]: nit: extra space

File: internal/pkg/cli/task_run.go (hunk 2 above)
Comment (Louise15926) [reply internal/pkg/cli/task_run.go:223]: That's true...My concern was that (assuming we want to have a short form for this flag because it is a required flag), if we use `--task-name` or `--task-family`, then it would probably make more sense to have `-t` as the short form. However, in app, svc, env, etc. it is always `-n`. Would it be considered inconsistent?

Comment (pkandasamy91) [reply internal/pkg/cli/task_run.go:223]: Ah, makes sense to keep it -n for consistency, but I think --task-name (--run-name or something along those lines) adds some clarity (that the task will be named as such, as well as the stack and all of it's resources?). Thoughts?

Comment (Louise15926) [reply internal/pkg/cli/task_run.go:223]: Ahh yes I agree. --name sounds like it is the name for a single task, instead of a group of similar tasks as intended. Probably something involving 'group' or 'family' might be good in order to emphasize that? --task-family-name as the long form and -n as the short form?

Comment (iamhopaul123) [reply internal/pkg/cli/task_run.go:223]: Actually `svc init` previously has the name flag called `--svc` instead of `--name`. And the reason why we made this change is because it seems to be redundant to have` svc init --svc frontend`, since `svc init --name frontend` already has enough background to know what `--name` refers to. I think maybe the same rule could be applied to `task run --name db-migrate` as well?

File: internal/pkg/cli/task_run.go [hunk 5]
```diff
@@ -129,6 +152,61 @@ func (o *runTaskOpts) validateEnvName() error {
...
+	if o.env != "" {
+		return nil
+	}
+
+	if o.AppName() == "" {
+		o.env = envNameNone
+		return nil
+	}
+
+	existingEnv, err := o.store.ListEnvironments(o.AppName())
```
Comment (bvtujo) [inline internal/pkg/cli/task_run.go:184]: Would it make sense to add a ListEnvironmentsWithNone method to the [ConfigSelector](https://github.com/aws/amazon-ecs-cli-v2/blob/0d26912c811d4c75d6746df2293fe753b315e911/internal/pkg/cli/selector/selector.go#L63) struct and wrap this complexity in an interface?

File: internal/pkg/cli/task_run.go (hunk 2 above)
Comment (bvtujo) [reply internal/pkg/cli/task_run.go:223]: I agree with using `--task-family-name/-n` for the name flag. We can use `--count / -c` for the number of tasks to run.

File: internal/pkg/cli/task_run.go [hunk 6]
```diff
@@ -110,6 +122,17 @@ func (o *runTaskOpts) Validate() error {
...
 }
 
+// Ask prompts the user for any required or important fields that are not provided.
+func (o *runTaskOpts) Ask() error {
+	if err := o.askTaskFamilyName(); err != nil {
+		return err
+	}
+	if err := o.askEnvName(); err != nil {
+		return err
+	}
```
Comment (bvtujo) [inline internal/pkg/cli/task_run.go:132]: Should we add an ask method for the number of tasks to run?

File: internal/pkg/cli/task_run.go (hunk 2 above)
Comment (pkandasamy91) [reply internal/pkg/cli/task_run.go:223]: --count / -c makes sense to me

File: internal/pkg/cli/task_run.go [hunk 7]
```diff
@@ -7,15 +7,26 @@ import (
...
+	errNumInvalid = errors.New("number of tasks must be positive and at most 10")
 	errCpuNotPositive = errors.New("CPU units must be positive")
 	errMemNotPositive = errors.New("memory must be positive")
+
+	fmtTaskRunEnvPrompt        = "Select an environment"
+	fmtTaskRunFamilyNamePrompt = fmt.Sprintf("What would you like to %s your task family?", color.Emphasize("name"))
+
+	taskRunEnvPromptHelp = fmt.Sprintf("Task will be deployed to the selected environment. " +
+		"Select %s to run the task in your default VPC instead of any existing environment.", color.Emphasize(envNameNone))
+	taskRunFamilyNamePromptHelp = "The family name of the task. Tasks with the same family name share the same set of resources."
```
[the remark this answers is not in the document]
Comment (pkandasamy91) [reply internal/pkg/cli/task_run.go:27]: Wouldn't this also refer to the stack name, the cw log group, etc? Does it make sense to call it task-family-name (it doesn't necessarily make sense in this context)?

File: internal/pkg/cli/task_run.go [hunk 8]
```diff
@@ -129,6 +152,61 @@ func (o *runTaskOpts) validateEnvName() error {
...
+	o.familyName = familyName
+	return nil
+}
+
+func (o *runTaskOpts) askEnvName() error {
+	if o.env != "" {
+		return nil
+	}
+
+	if o.AppName() == "" {
```
[the remark this answers is not in the document]
Comment (Louise15926) [reply internal/pkg/cli/task_run.go:179]: AppName() could still be "" here when the `task run` command isn't run under a workspace and the user didn't provide an app name. Unlike `svc init`, `task run` doesn't have to be under a workspace, in which case we will just use the default VPC and cluster instead of an environment previously created by Copilot (hence the envNameNone)

File: internal/pkg/cli/task_run.go (hunk 5 above)
Comment (Louise15926) [reply internal/pkg/cli/task_run.go:184]: I didn't do it because `ListEnvironmentsWithNone` wouldn't get used anywhere else. It'd also definitely make sense though to wrap this in an interface. In that case, would it make sense to add that method to [Select](https://github.com/aws/amazon-ecs-cli-v2/blob/0d26912c811d4c75d6746df2293fe753b315e911/internal/pkg/cli/selector/selector.go#L37)  ?

File: internal/pkg/cli/task_run.go (hunk 6 above)
Comment (Louise15926) [reply internal/pkg/cli/task_run.go:132]: That would basically make "number of tasks" a required field right? But it is (currently meant to be) an optional field which default value to 1.

Comment (iamhopaul123) [reply internal/pkg/cli/task_run.go:132]: We can have default value for the prompt but +1 on that, since I think for most of cases the num will just be 1.

File: internal/pkg/cli/validate.go [hunk 9]
```diff
@@ -120,6 +120,13 @@ func validateEnvironmentName(val interface{}) error {
 	return nil
 }
 
+func validateTaskFamilyName(val interface{}) error {
```
[the remark this answers is not in the document]
Comment (Louise15926) [reply internal/pkg/cli/validate.go:123]: That totally makes sense. My concern was that unwrapping it would be inconsistent with what `svc` and `env` is doing (there are `validateSvcName` and `validateEnvironmentName` and they are called at similar places like `validateTaskFamilyName` i.e. at `Validate()` and `o.prompt.Get()`). Any thought?

File: internal/pkg/cli/task_run.go (hunk 2 above)
Comment (Louise15926) [reply internal/pkg/cli/task_run.go:223]: I agree with using `--count / -c`!

I also agree with @pkandasamy91 on that `--task-family-name` could be potentially confusing. The ECS task definition has a `family` property, while we don't mean exactly the same thing here since it also involves cw log group, ecr repo, etc. 

I am also proposing `--task-group-name` because it not only indicates that multiple tasks can be "grouped" together, but also is different than the "family" property in ECS task definition.

File: internal/pkg/cli/task_run.go (hunk 5 above)
Comment (bvtujo) [reply internal/pkg/cli/task_run.go:184]: Yes, I would add that to Select as you say. And I'd create a [new interface](https://github.com/aws/amazon-ecs-cli-v2/blob/0d26912c811d4c75d6746df2293fe753b315e911/internal/pkg/cli/interfaces.go#L335) `appEnvWithNoneSelector` specifying `configSelector` and `EnvironmentsWithNone()`.

This will let you mock calls to this method and hopefully simplify your testing.

File: internal/pkg/cli/validate.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply internal/pkg/cli/validate.go:123]: That's a good call. What we currently have for `validateSvcName` is actually a bad behavior. Ideally we should only wrap errors for cross package methods and this error should be something like `task definition %s is not allowed because of %w` as a single layer of error. The same applies to the other similar methods but might need more refactoring work to validate.go.

File: internal/pkg/cli/selector/selector.go [hunk 10]
```diff
@@ -140,6 +140,28 @@ func (s *Select) Environment(prompt, help, app string) (string, error) {
 	return selectedEnvName, nil
 }
 
+// EnvironmentWithNone fetches all the environments in an app and prompts the user to select one of the environments or None.
+func (s *Select) EnvironmentWithNone(prompt, help, app string) (string, error) {
+	envs, err := s.retrieveEnvironments(app)
+	if err != nil {
+		return "", fmt.Errorf("get environments for app %s from metadata store: %w", app, err)
```
Comment (bvtujo) [inline internal/pkg/cli/selector/selector.go:147]: Can we add a unit test that will hit this error block?

File: internal/pkg/cli/task_run.go [hunk 11]
```diff
@@ -124,7 +124,7 @@ func (o *runTaskOpts) Validate() error {
 
 // Ask prompts the user for any required or important fields that are not provided.
 func (o *runTaskOpts) Ask() error {
-	if err := o.askTaskFamilyName(); err != nil {
+	if err := o.askTaskGroupName(); err != nil {
 		return err
 	}
 	if err := o.askEnvName(); err != nil {
```
Comment (bvtujo) [inline internal/pkg/cli/task_run.go:130]: Can we add a unit test to mock this failure and ensure that the error is wrapped?

Comment (bvtujo) [reply internal/pkg/cli/task_run.go:130]: Also on this method call, can we add a unit test which returns "None" to ensure that everything works as expected in the case that we have to check the default cluster?

File: internal/pkg/cli/selector/selector.go (hunk 10 above)
Comment (Louise15926) [reply internal/pkg/cli/selector/selector.go:147]: Good call..! I realized that `EnvironmentWithNone` and `Environment` wrap the error returned from `retrieveEnvironments`, which is from the same package. Is this wrapping necessary? I feel like I can just return "", err for line 147.

Comment (bvtujo) [reply internal/pkg/cli/selector/selector.go]: I think this error can be passed through like you say instead of being wrapped.
